README.md
# Competitive Refresh

<h4 align="center">Be sure to :star: this repository so you can keep up to date on any updates!</h4>
<p align="center">
 <img src ='https://forthebadge.com/images/badges/made-with-python.svg'>
 <img src ='https://forthebadge.com/images/badges/open-source.svg'>
</p>

## 📑 Table of Contents 📑
 - [Description](https://github.com/Accenture-Intel/competitive_refresh/edit/main/README.md#description)
 - [Accenture - Intel Partnership](https://github.com/Accenture-Intel/competitive_refresh/edit/main/README.md#accenture---intel-partnership)
 - [Benchmark Details](https://github.com/Accenture-Intel/competitive_refresh/edit/main/README.md#benchmark-details)
 - [Environment Setup](https://github.com/Accenture-Intel/competitive_refresh/edit/main/README.md#environment-setup)
 - [Benchmark-Specific Code Deployment and Results](https://github.com/Accenture-Intel/competitive_refresh/edit/main/README.md#benchmark-specific-code-deployment-and-results)

## Description

As part of an Accenture-Intel partnership aimed at accelerating client transformation and driving co-innovation through cutting-edge technology and industry experience, we are showcasing performance and price-performance metrics across AWS instances on a variety of AI/ML use-cases. We invite you to explore our workloads and build upon them using Intel's platform.

## Accenture - Intel Partnership
What do you get when you combine a company that “delivers on the promise of technology and human ingenuity” and another that “shapes the future of technology”?
Accelerated co-innovation; capabilities alignment: the right experts at the right time; and consistent client outcomes thanks to trained Intel solution architects at Accenture. 

Renowned as titans in the high-tech space, Accenture and Intel share a single-minded focus to accelerate client transformation and drive co-innovation. The Accenture and Intel joint partnership bring together cutting-edge technology and unrivaled market & industry expertise allowing for significant expansion into the following portfolio offerings (towers): Multicloud, Network, Analytics and AI, and Digital Workspace. Each tower provides generates significant value not only to both Accenture and Intel, but also to their clients. These towers show the co-dependency Accenture and Intel have, with Intel bringing optimized AI offerings using its Xeon and adjacent technologies and Accenture bringing its Applied Intelligence Platform (AIP) & Solutions.AI offerings for intelligent revenue growth and advanced consumer engagement, for example. Together, Accenture and Intel push industries beyond their limits, providing consistent outcomes and evangelizing success for the benefit of their clients.

To learn more about this partnership, follow this <a href="https://www.accenture.com/us-en/services/high-tech/alliance-intel/" target="_blank">link</a>.

## Benchmark Details  
### Purpose
In order to update Intel processor positioning relative to competitors' most recent AWS instance offerings, we compared the performance of the Intel Ice Lake m6i against the AMD EPYC m6a and the AWS Graviton2 t4g & m6g. Using four use-cases meant to test commonly-used models and algorithms, we analyzed each instance's prediction time to determine a performance and price-to-performance metric relative to eachother, comparing both the stock model and the Intel oneAPI speedup accessible on the Intel and AMD instances.

<img width="857" alt="Screenshot 2022-08-09 111626" src="https://user-images.githubusercontent.com/107268111/183730984-5bbc3167-e237-43de-acbc-a5c9fb33226e.png">

## Environment Setup

### AWS Instance Launch

#### m6i and m6a EC2 Instances

Please follow the the instructions given below for setting up the environment. The guide will walk you through the process of setting up your system to run the model on the m6i EC2 instance. The procedure is the same for the m6a instance, just select the **m6a instance** from the _All instance families_ drop-down at the end of the setup.

##### Launching the Deep Learning AMI (DLAMI)
To launch the DLAMI, go to your AWS EC2 page and click 'Launch instances'.
![image](https://user-images.githubusercontent.com/91902558/157768843-a3a73db5-9e01-45c2-ac0b-285fa11d6c46.png)

In the searchbar, search 'Deep Learning AMI' and select AWS Marketplace. From the options, select the **Deep Learning AMI (Ubuntu 18.04) Version 57.0** option and press continue on the prompt.
![image](https://user-images.githubusercontent.com/91902558/157768743-dc568c48-5cc6-4951-b81d-ba24d6f6db55.png)

After pressing continue on the prompt, select the **m6i** instance from the _All instance families_ drop-down as recommended for this AMI and configure the instance details as needed. Click on the **Review and Launch** and **Launch** buttons at the bottom right corner of your screen, and follow the steps for key pairs.
![image](https://user-images.githubusercontent.com/91902558/157769116-01fc2a0a-4846-479f-b65f-2fe75df3468e.png)

Once all of that is complete, you can launch & connect to the Deep Learning AMI (Ubuntu 18.04) Version 57.0.

#### t4g and m6g EC2 Instances

The Graviton2 instances run on an ARM-based processor rather than x86 like Intel and AMD processors; as a result, the setup for these instances is slightly different from above. When searching for the AMI, instead of searching for the 'Deep Learning AMI', search for **Deep Learning AMI Graviton** and, from the *Community AMIs* section, select **Deep Learning AMI Graviton TensorFlow 2.7.0 (Ubuntu 20.04) 20220107**.

<img width="970" alt="Screenshot 2022-08-08 145200" src="https://user-images.githubusercontent.com/107268111/183521349-f17bae9f-ba79-426d-8b2c-1da7c50e7ab3.png">

### Instance Setup
Different instances require different instances and setup before running the benchmarks. Please run the below commands in your terminal after ssh-ing into the appropriate EC2 instance to setup packages and clone into the repo.

#### Intel Ice Lake m6i and AMD EPYC m6a
```
- git clone https://github.com/Accenture-Intel/competitive_refresh
```

#### AWS Graviton2 t4g & m6g
```
- curl https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-aarch64.sh -o Miniconda3-latest-Linux-aarch64.sh
- bash Miniconda3-latest-Linux-aarch64.sh
*restart your EC2 instance*
- conda create -n tf_bench python=3.8
- conda activate tf_bench
- conda install -c conda-forge tensorflow jupyter pandas
- pip install sklearn tensorflow-aarch64
- export LD_PRELOAD=/usr/lib/aarch64-linux-gnu-libgomp.so.1
- git clone https://github.com/Accenture-Intel/competitive_refresh
```

## Benchmark-Specific Code Deployment and Results

### Food Demand Forecasting - XGBoost Machine Learning

#### Description and Dataset
This benchmark uses an XGBoost regression model to predict demand for a meal delivery company. Given a dataset with 15 features and over 500,000 meal deliveries over 145 weeks, the goal is to predict the demand for various raw materials for the following 10 weeks. [Here](https://www.kaggle.com/datasets/kannanaikkal/food-demand-forecasting) is a link to the dataset for more information, but the CSVs are already in the repository and don't need to be downloaded. To learn more about the XGBoost algorithm, please check out our [XGBoost repo](https://github.com/Accenture-Intel/XGBoost#description-1).

#### Running the Benchmark
Once inside the xgboost folder, open a Jupyter Notebook and run the **food_demand_forecasting.ipynb** notebook file.

#### Results
Below are two graphs of the results of our tests. Both compare the relative prediction time improvements of each instance and software compared to the Intel Ice Lake m6i instance running stock XGBoost. The left graph compares pure prediction performance gain, and the right one compares the same performance gain but relative to the per-unit cost of each of the instances. In both cases, **Ice Lake with Intel daal4py optimizations have the highest metrics** across the tests. 

<img width="875" alt="Screenshot 2022-08-09 103302" src="https://user-images.githubusercontent.com/107268111/183721289-384b4185-5a26-48d1-aa92-c5528235635d.png">

### Chest X-Ray Pneumonia Detection - CNN Deep Learning

#### Description and Dataset
This benchmark uses an Convolutional Neural Network to predict whether a given chest X-ray shows that the patient has pneumonia. The dataset of 5,863 images of normal and pneumonic patients has already been downloaded in the repository, but can be found [here](https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia) for more information. For more information about CNN, please check out our [CNN Hyperparameter Tuning repo](https://github.com/Accenture-Intel/Hyperparameter_Tuning/tree/main/cnn#description-1).

#### Running the Benchmark
If on an m6i or m6a instance, open Jupyter Notebook and run the **cnn_pneumonia.ipynb** inside the **conda_tensorflow2_p39** environment, which can be accessed through the *New* dropdown on the top right of the screen.

<img width="280" alt="Screenshot 2022-08-08 155854" src="https://user-images.githubusercontent.com/107268111/183530071-6ae71cf9-d961-478d-bf9f-fd01144b57b5.png">

If on a t4g or m6g instance, simply run the **cnn_pneumonia.py** script from the **tf_bench** conda environment in your terminal.

#### Results
Below are two graphs of the results of our tests. Both compare the relative prediction time improvements of each instance and software compared to the Intel Ice Lake m6i instance. The Ice Lake and EPYC instances are running built-in Intel oneDNN software optimizations on top of the Tensorflow neural network package, and the Graviton2 instances are running stock Tensorflow. The left graph compares pure prediction performance gain, and the right one compares the same performance gain but relative to the per-unit cost of each of the instances. In both cases, **Ice Lake with Intel oneDNN optimizations has the highest metrics** across the tests. 

<img width="875" alt="Screenshot 2022-08-09 104925" src="https://user-images.githubusercontent.com/107268111/183724320-a0026312-16cc-4a56-8152-caade48d7661.png">

### Mortgage Loan Prediction - Random Forest Regression

#### Description and Dataset
This benchmark uses a Random Forest regression model to predict the optimal mortgage loan amount to lend to a customer to minimize default risk. The dataset of over 2 billion rows and 108 columns should be downloaded from [this](https://capitalmarkets.fanniemae.com/credit-risk-transfer/single-family-credit-risk-transfer/fannie-mae-single-family-loan-performance-data) link. Doublecheck that the downloaded ZIP file's filename is exactly **ML_2020Q3Mortgage.zip**, and that the **ML_2020Q3Mortgage.csv** file that is extracted from that is in the same random_forest folder as the .py and .ipynb scripts. Alternatively, if you don't want to download the dataset you can use the ZIP file in the repo that contains a truncated version of the CSV with only 80% of the rows; if you use this, make sure that the filepath and name of the extracted CSV is the same as above. For more information about random forests, please check out this [link](https://towardsdatascience.com/random-forest-explained-7eae084f3ebe#:~:text=In%20the%20Machine%20Learning%20worl%20d%2C%20Random%20Forest,belonging%20to%20the%20specific%20category%20of%20Bagging%20methods.).

#### Running the Benchmark
If on an m6i or m6a instance, run the **mortgage_prediction_daal_and_stock.py** script in your terminal (alternatively you can also run the .ipynb). If on a t4g or m6g, run the **mortgage_prediction_stock_only.py** script, because the Intel OneAPI daal4py software optimizer isn't compatible with ARM-based instances.

#### Results
Below are two graphs of the results of our tests. Both compare the relative prediction time improvements of each instance and software compared to the Intel Ice Lake m6i instance running stock scikit-learn Random Forests. The left graph compares pure prediction performance gain, and the right one compares the same performance gain but relative to the per-unit cost of each of the instances. In both cases, **Ice Lake with Intel daal4py optimizations have the highest metrics** across the tests. 

<img width="887" alt="Screenshot 2022-08-09 111305" src="https://user-images.githubusercontent.com/107268111/183729388-428ccdff-a339-47ce-b9eb-66e3fa040a0d.png">

### CoLA Grammatical Acceptability - BERT NLP Deep Learning

#### Description and Dataset
This benchmark uses a BERT neural network to judge the grammatical acceptability of a sentence. This Corpus of Linguistic Acceptability (CoLA) dataset is a set of ~10,000 English sentences that have been marked as grammatical or ungrammatical from published linguistics literature, and is a common test used to determine how effective a model is at Natural Language Processing along with other [GLUE Benchmarks](https://gluebenchmark.com). This dataset will be loaded from the script you run, and nothing needs to be downloaded beforehand. For more information about BERT and NLP, please check out our [Fine-Tune BERT for Text Classification repo](https://github.com/Accenture-Intel/Fine-Tune-BERT-for-Text-Classification#bert-for-text-classification).

#### Running the Benchmark
Before running the test, please run the following command in your terminal:
```
pip install datasets evaluate transformers[sentencepiece]
```

There are two options for how to run this test. You could either run the **bert_cola.ipynb** notebook in a Jupyter Notebook, orthe **bert_cola.py** script in your terminal. If you're doing this on an Intel or AMD instance, make sure that you either run the .ipynb from the same **conda_tensorflow2_p39** environment used in the CNN X-Ray test above, or you first type ```source activate tensorflow2_p39``` into your terminal before running the script to be sure you have all the optimized libraries activated

#### Results
Below are two graphs of the results of our tests. Both compare the relative prediction time improvements of each instance and software compared to the Intel Ice Lake m6i instance. The left graph compares pure prediction performance gain, and the right one compares the same performance gain but relative to the per-unit cost of each of the instances. In both cases, **Ice Lake with Intel oneDNN optimizations has the highest metrics** across the tests. 

<img width="881" alt="Screenshot 2022-08-11 130401" src="https://user-images.githubusercontent.com/107268111/184231351-8dc7ced5-b6ae-46f8-a63f-1d9e5bf24f74.png">

## Used By

This project is used by the following companies:
![ACN-Intel_logo](https://user-images.githubusercontent.com/91902558/157770015-ea092843-c4ee-4fb4-8207-31c0368d718b.png)


splunk-uf-installer-linux-x86_64-9.1.0.1r1/changelog
9.1.0.1r1	20230804
	* New Splunk release only
9.0.5r1		20230608
	* Installer:
		- version 3.16
		- Added INSTALL_METHOD to track type of installer
		- remove any orphan init
		- Image installation mode
	* Uninstaller:
		- version 1.7
		- Make sure SPLUNK_HOME path is defined
		- improved disable boot-start
9.0.4r4		20230519
	* 000_accenture_ds:
		- updated certificates in CA bundle digicert.ca.bundle.pem
	* Installer:
		- version 3.15
		- Improvements to Splunk init
9.0.4r3		20230516
	* Installer:
		- version 3.14
		- Improvements to Splunk init
9.0.4r2		20230424
	* Installer:
		- version 3.13
		- Option to define HF type
	* Uninstaller:
		- version 1.6
		- Switch -y to skip validation of uninstallation
9.0.4r1		20230330
	* Installer:
		- version 3.12
		- Minor updates to ACNIS files
9.0.3r1		20230210
	* Installer:
		- version 3.11
		- Minor updates to better identify Splunk ACNIS installations
	* 000_accenture_linux_is:
		- app version 2.1.6
		- updated from DS tree
9.0.1r1		20221027
	* Installer:
		- version 3.10
		- Fix for systems using cgroups v2 without system.slices
		- Splunk init improved
9.0.0r3		20220628
	* 000_accenture_linux_is:
		- app version 2.1.2
		- get_env_details.sh - openssl call fix
9.0.0r2		20220620
	* 000_accenture_linux_is:
		- app version 2.1.1
		- get_audit_ng.sh - improved processing for smaller mem footprint
9.0.0r1		20220615
	* Installer:
		- version 3.9
		- add clientName prefix to DS client config when it's HF (Type D or V&A)
	* 000_accenture_linux_is:
		- app version 2.1.0
		- adds get_audit_ng.sh script to replace python audit2splunk.py parser
	* Splunk_TA_nix:
		- app version 8.5.0
8.1.0.1r21	20220110
	* Installer:
		- version 3.8
		- do not alter local/server.conf when 000_accenture_ds app is not installed
8.1.0.1r20	20211216
	* 000_accenture_linux_is:
		- add_meta.sh - version 1.6
		- add_meta.sh - added HF meta
	* Installer:
		- version 3.7
		- UF and HF installer unified
		- external config file to set instance type and architecture
		- switch -t merged from original HF installer to define site id (HF only)
	* Uninstaller
		- version 1.5
		- modified to work with both UF and HF
8.1.0.1r19	20211201
	* 000_accenture_linux_is:
		- get_audit.sh - version 2.0
		- get_audit.sh - handles all the logic for collecting events using ausearch
		- get_audit.sh - pass events to python processor via stdout in pipe
		- get_audit.sh - uses bash processor (w/o filtering) if audit.log is larger than 100 MB
		- get_audit.sh - only proceses rotated audit.log.X files when they are less than 10 MB of size
		- audit2splunk.py - version 3.0
		- audit2splunk.py - removed logic to handle ausearch
		- audit2splunk.py - reads events from stdin
8.1.0.1r18	20211111
	* Installer: fixed syntax error
8.1.0.1r17	20211108
	* 000_accenture_linux_is:
		- get_audit.sh - version 1.3
		- get_audit.sh - improved logging, checks ausearch availability
		- audit2splunk.py - version 2.5
		- audit2splunk.py - improved logging
	* 000_accenture_ds:
		- deploymentclient.conf - call home set to 600 seconds
8.1.0.1r16	20211015
	* 000_accenture_linux_is:
		- get_env_details.sh - ver 3.3, adds journald details, adds Splunk version, improves OS detection
		- inputs.conf uses Python Audit scripted input by default for audit events, uses JournalD by default for system events
	* Installer:
		- Keep defaul settings of Python Auditd scripted input, switch to /var/log monitoring only if JournalD is not available
8.1.0.1r15	20210826
	* 000_accenture_linux_is:
		- inpust.conf - get_audit.sh scripted input as default for collection of audit logs
		- get_audit.sh - Reads events directly if no Python interpreter found, but without option to filter out unwanted events
		- get_env_details.sh - accountId field is collected from cloud platforms
	* 000_accenture_ds:
		- new app to integrate agent with IS Deployment Server
	* Installer:
		- added code to backup system/local/server.conf and comment out sslPassword from it to use server.conf values from 000_accenture_ds app
		- DS config status added to install log event
8.1.0.1r14	20210811
	* 000_accenture_linux_is:
		- inpust.conf - /var/log/audit/audit.log set as default
	* Installer:
		- do not set get_audit.sh as default
8.1.0.1r13	20210628
	* 000_accenture_linux_is:
		- audit2splunk.py version 2.4
		- audit2splunk.py do not read audit.log.x files without permissions
		- audit2splunk.py make sure list of event IDs to exclude is unique
8.1.0.1r12	20210618
	* 000_accenture_linux_is:
		- audit2splunk.py version 2.3
		- audit2splunk.py fixed reading of config file because of relative paths
		- add_meta.sh version 1.5
		- add_meta.sh prevents UF restart loop
		- get_env_details.sh version 3.0
		- get_env_details.sh Fixed malformed output from python version check
	* audit.rules:
		- version 1.8
		- Use -i instead of -c to skip unkonwn rules to do not break MS ATP
8.1.0.1r12b3	20210525
	* 000_accenture_linux_is:
		- audit2splunk.py version 2.2
		- audit2splunk.py allows multiple events of the same type under single event_id
		- audit2splunk.py deals with unparsed events
	* Installer:
		- version 3.1
		- installs audit rules in older system without rules.d
8.1.0.1r12b2	20210525
	* 000_accenture_linux_is:
		- get_env_details.sh version 2.9
		- get_env_details.sh checks Python availability & version
		- audit2splunk.py version 2.1
		- audit2splunk.py improved config parsing
		- audit2splunk.py merges config from $app$/default/audit2splunk.conf and $app$/local/audit2splunk.conf
	* audit.rules:
		- version 1.7
		- Re-enabled rule acnis-command-executed-by-user with filtering in audit2splunk.py
8.1.0.1r12b1	20210524
	* Installer:
		- version 3.0
		- Python as optional dependency
		- cleans up output app using old naming convetion
	* 000_accenture_linux_is:
		- audit2splunk.py & get_audit.sh version 2.0
		- audit2splunk.py first release in beta package
		- audit2splunk.py reads patterns from $app$/local/audit2splunk.conf

cnn/cnn_pneumonia.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import os
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import time 


# In[2]:


print(tf.__version__)


# In[3]:


AUTOTUNE = tf.data.experimental.AUTOTUNE
BATCH_SIZE = 128 
IMAGE_SIZE = [180, 180]
EPOCHS = 25


# In[4]:


filenames = tf.io.gfile.glob(str('train/*/*'))
filenames.extend(tf.io.gfile.glob(str('val/*/*')))

train_filenames, val_filenames = train_test_split(filenames, test_size=0.2)


# In[5]:


# import pickle

# pickle.dump(train_filenames, open('train_filenames.pkl', 'wb'))
# pickle.dump(val_filenames, open('val_filenames.pkl', 'wb'))


# In[6]:


len(train_filenames), len(val_filenames)


# In[7]:


COUNT_NORMAL = len([filename for filename in train_filenames if "NORMAL" in filename])
print("Normal images count in training set: " + str(COUNT_NORMAL))

COUNT_PNEUMONIA = len([filename for filename in train_filenames if "PNEUMONIA" in filename])
print("Pneumonia images count in training set: " + str(COUNT_PNEUMONIA))


# In[8]:


train_list_ds = tf.data.Dataset.from_tensor_slices(train_filenames)
val_list_ds = tf.data.Dataset.from_tensor_slices(val_filenames)

for f in train_list_ds.take(5):
    print(f.numpy())


# In[9]:


TRAIN_IMG_COUNT = tf.data.experimental.cardinality(train_list_ds).numpy()
print("Training images count: " + str(TRAIN_IMG_COUNT))

VAL_IMG_COUNT = tf.data.experimental.cardinality(val_list_ds).numpy()
print("Validating images count: " + str(VAL_IMG_COUNT))


# In[10]:


def get_label(file_path):
    # convert the path to a list of path components
    parts = tf.strings.split(file_path, os.path.sep)
    # The second to last is the class-directory
    return parts[-2] == "PNEUMONIA"


# In[11]:


def decode_img(img):
  # convert the compressed string to a 3D uint8 tensor
  img = tf.image.decode_jpeg(img, channels=3)
  # Use `convert_image_dtype` to convert to floats in the [0,1] range.
  img = tf.image.convert_image_dtype(img, tf.float32)
  # resize the image to the desired size.
  return tf.image.resize(img, IMAGE_SIZE)


# In[12]:


def process_path(file_path):
    label = get_label(file_path)
    # load the raw data from the file as a string
    img = tf.io.read_file(file_path)
    img = decode_img(img)
    return img, label


# In[13]:


train_ds = train_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)

val_ds = val_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)


# In[14]:


for image, label in train_ds.take(1):
    print("Image shape: ", image.numpy().shape)
    print("Label: ", label.numpy())


# In[15]:


test_list_ds = tf.data.Dataset.list_files(str('test/*/*'))
TEST_IMAGE_COUNT = tf.data.experimental.cardinality(test_list_ds).numpy()
test_ds = test_list_ds.map(process_path, num_parallel_calls=AUTOTUNE)
test_ds = test_ds.batch(BATCH_SIZE)

TEST_IMAGE_COUNT


# In[16]:


def prepare_for_training(ds, cache=True, shuffle_buffer_size=1000):
    # This is a small dataset, only load it once, and keep it in memory.
    # use `.cache(filename)` to cache preprocessing work for datasets that don't
    # fit in memory.
    if cache:
        if isinstance(cache, str):
            ds = ds.cache(cache)
        else:
            ds = ds.cache()

    ds = ds.shuffle(buffer_size=shuffle_buffer_size)

    # Repeat forever
    ds = ds.repeat()

    ds = ds.batch(BATCH_SIZE)

    # `prefetch` lets the dataset fetch batches in the background while the model
    # is training.
    ds = ds.prefetch(buffer_size=AUTOTUNE)

    return ds


# In[17]:


train_ds = prepare_for_training(train_ds)
val_ds = prepare_for_training(val_ds)

image_batch, label_batch = next(iter(train_ds))


# In[18]:


def conv_block(filters):
    block = tf.keras.Sequential([
        tf.keras.layers.SeparableConv2D(filters, 3, activation='relu', padding='same'),
        tf.keras.layers.SeparableConv2D(filters, 3, activation='relu', padding='same'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPool2D()
    ]
    )
    
    return block


# In[19]:


def dense_block(units, dropout_rate):
    block = tf.keras.Sequential([
        tf.keras.layers.Dense(units, activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate)
    ])
    
    return block


# In[20]:


def build_model():
    model = tf.keras.Sequential([
        tf.keras.Input(shape=(IMAGE_SIZE[0], IMAGE_SIZE[1], 3)),
        
        tf.keras.layers.Conv2D(16, 3, activation='relu', padding='same'),
        tf.keras.layers.Conv2D(16, 3, activation='relu', padding='same'),
        tf.keras.layers.MaxPool2D(),
        
        conv_block(32),
        conv_block(64),
        
        conv_block(128),
        tf.keras.layers.Dropout(0.2),
        
        conv_block(256),
        tf.keras.layers.Dropout(0.2),
        
        tf.keras.layers.Flatten(),
        dense_block(512, 0.7),
        dense_block(128, 0.5),
        dense_block(64, 0.3),
        
        tf.keras.layers.Dense(1, activation='sigmoid')
    ])
    
    return model


# In[21]:


initial_bias = np.log([COUNT_PNEUMONIA/COUNT_NORMAL])
initial_bias


# In[22]:


weight_for_0 = (1 / COUNT_NORMAL)*(TRAIN_IMG_COUNT)/2.0 
weight_for_1 = (1 / COUNT_PNEUMONIA)*(TRAIN_IMG_COUNT)/2.0

class_weight = {0: weight_for_0, 1: weight_for_1}

print('Weight for class 0: {:.2f}'.format(weight_for_0))
print('Weight for class 1: {:.2f}'.format(weight_for_1))


# In[23]:


model = build_model()

METRICS = [
    'accuracy',
    tf.keras.metrics.Precision(name='precision'),
    tf.keras.metrics.Recall(name='recall')
]

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=METRICS
)


# In[24]:


start = time.time()
history = model.fit(
    train_ds,
    steps_per_epoch=TRAIN_IMG_COUNT // BATCH_SIZE,
    epochs=EPOCHS,
    validation_data=val_ds,
    validation_steps=VAL_IMG_COUNT // BATCH_SIZE,
    class_weight=class_weight,
)
end = time.time() - start
print("total training time: ", end)


# In[25]:


print('Training time (mins): {}'.format(end/60))


# In[26]:


start_eval = time.time()

loss, acc, prec, rec = model.evaluate(test_ds)

end_eval = time.time() - start_eval

print("loss: ", loss)

print("acc: ", acc)

print("prec: ", prec)

print("rec: ", rec)

print("total prediction time: ", end_eval)



random_forest/mortgage_prediction_daal_and_stock.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


get_ipython().system('pip install scikit-learn-intelex')


# In[ ]:


get_ipython().system('unzip ML_2020Q3Mortage.zip')


# In[ ]:


# organize notebook format
# To expand output so that it shows all data columns 
import pandas as pd
import daal4py as d4p
import numpy as np
import time
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

# to hide warnings
import warnings
warnings.filterwarnings('ignore')


# In[ ]:


df_ml = pd.read_csv('ML_2020Q3Mortage.csv')


# In[ ]:


df_ml


# In[ ]:


#10% of full dataset to test daal model 
# train_dataset = df_ml.iloc[:277199,:]


# In[ ]:


from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error


# In[ ]:


# split data to train/test
X = df_ml.drop(['LOAN_ID','ORIG_UPB'], axis=1)
y = df_ml.ORIG_UPB

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)


# In[ ]:


# delete unused further data to free up GPU memory
del df_ml , X ,y


# In[ ]:


d4p.daalinit()


# In[ ]:


model = d4p.decision_forest_regression_training(nTrees=100)


# In[ ]:


# model 
start_stock = time.time()
reg = RandomForestRegressor().fit(X_train, y_train)
end_stock = time.time() - start_stock
print("Stock training time: ", end_stock)


# In[ ]:


start_pred = time.time()
reg.predict(X_test)
end_pred = time.time() - start_pred
print('Stock pred time: ', end_pred)


# In[ ]:


# optimized model 
start = time.time()
train_result = model.compute(X_train, y_train)
end = time.time() - start
print("Intel optimized training time: ", end)


# In[ ]:


start_pred = time.time()
pred_algo = d4p.decision_forest_regression_prediction(fptype='float')
predict_res = pred_algo.compute(X_test, train_result.model)
end_pred = time.time() - start_pred
print('Intel pred time: ', end_pred)



splunk-uf-installer-linux-x86_64-9.1.0.1r1/collect_logs.sh
#!/bin/bash
# This script can be used to collect basic data about source system when Splunk installation fails.
# It can help Splunk Team to identify what could be the issue.
# Version 1.1

LOG="splunk-debug.log"

echo -n "Collecting logs and environment details..."

date > $LOG

echo "validating /etc/passwd" >> $LOG
grep splunk /etc/passwd  >> $LOG 2>&1
echo "---" >> $LOG

echo "validating /etc/group" >> $LOG
grep splunk /etc/group  >> $LOG 2>&1
echo "---" >> $LOG

echo "checking SELinux" >> $LOG
getenforce >> $LOG 2>&1
echo "---" >> $LOG

echo "checking disks" >> $LOG
df -h  >> $LOG 2>&1
echo "---" >> $LOG

echo "search for read-only filesystems" >> $LOG
grep "[[:space:]]ro[[:space:],]" /proc/mounts >> $LOG 2>&1
echo "---" >> $LOG

echo "checking mem" >> $LOG
free -h  >> $LOG 2>&1
echo "---" >> $LOG

echo "check permissions of /opt" >> $LOG
ll / | grep opt  >> $LOG 2>&1
echo "---" >> $LOG

echo "getting processes" >> $LOG
ps aux  >> $LOG 2>&1
echo "---" >> $LOG

echo "getting list of RPM packages" >> $LOG
rpm -qa >> $LOG 2>&1
echo "---" >> $LOG

echo "getting list of DEB packages" >> $LOG
dpkg -l >> $LOG 2>&1
echo "---" >> $LOG

echo "getting recent events from journald" >> $LOG
ausearch --start today >> $LOG 2>&1
echo "---" >> $LOG

echo "getting kernel dmesg" >> $LOG
dmesg >> $LOG 2>&1
echo "---" >> $LOG

DEBUGARCHIVE="splunk-$HOSTNAME-debug.tar.gz"

tar chfz $DEBUGARCHIVE install-err.log splunk-debug.log /etc/audit /etc/ld.so.* /etc/os-release /etc/security /etc/selinux /etc/sysconfig /opt/splunkforwarder/etc /opt/splunkforwarder/var/log /var/log/audit/* /var/log/messages /var/log/syslog 2>/dev/null 1>/dev/null

echo "done."

echo "Please share $DEBUGARCHIVE with Splunk Team."


bert/bert_cola.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


get_ipython().system('pip install datasets evaluate transformers[sentencepiece]')


# In[16]:


from datasets import load_dataset
from transformers import AutoTokenizer, DataCollatorWithPadding
import numpy as np
import tensorflow as tf
import time


# In[13]:


print(tf.__version__)


# In[2]:


task = "cola"
dataset = load_dataset("glue", task)
checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)


# In[14]:


dataset


# In[3]:


task_to_keys = {
    "cola": ("sentence", None),
    "mnli": ("premise", "hypothesis"),
    "mnli-mm": ("premise", "hypothesis"),
    "mrpc": ("sentence1", "sentence2"),
    "qnli": ("question", "sentence"),
    "qqp": ("question1", "question2"),
    "rte": ("sentence1", "sentence2"),
    "sst2": ("sentence", None),
    "stsb": ("sentence1", "sentence2"),
    "wnli": ("sentence1", "sentence2"),
}


# In[4]:


sentence1_key, sentence2_key = task_to_keys[task]


# In[5]:


def tokenize_function(examples):
    if sentence2_key is None:
        return tokenizer(examples[sentence1_key], truncation=True)
    return tokenizer(examples[sentence1_key], examples[sentence2_key], truncation=True)


# In[7]:


tokenized_datasets = dataset.map(tokenize_function, batched=True)


# In[8]:


data_collator = DataCollatorWithPadding(tokenizer=tokenizer, return_tensors="tf")


# In[9]:


tf_train_dataset = tokenized_datasets["train"].to_tf_dataset(
    columns=["attention_mask", "input_ids", "token_type_ids"],
    label_cols=["labels"],
    shuffle=True,
    collate_fn=data_collator,
    batch_size=8,
)

tf_validation_dataset = tokenized_datasets["validation"].to_tf_dataset(
    columns=["attention_mask", "input_ids", "token_type_ids"],
    label_cols=["labels"],
    shuffle=False,
    collate_fn=data_collator,
    batch_size=8,
)


# In[10]:


from transformers import TFAutoModelForSequenceClassification

model = TFAutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)


# In[24]:


from tensorflow.keras.losses import SparseCategoricalCrossentropy

model.compile(
    optimizer="adam",
    loss=SparseCategoricalCrossentropy(from_logits=True),
    metrics=["accuracy"],
)
start = time.time()
model.fit(
    tf_train_dataset,
    validation_data=tf_validation_dataset,
)
end = time.time() - start
print("Training Time: ", end)


# # Fine Tuning 

# In[25]:


from tensorflow.keras.optimizers.schedules import PolynomialDecay

batch_size = 8
num_epochs = 3
# The number of training steps is the number of samples in the dataset, divided by the batch size then multiplied
# by the total number of epochs. Note that the tf_train_dataset here is a batched tf.data.Dataset,
# not the original Hugging Face Dataset, so its len() is already num_samples // batch_size.
num_train_steps = len(tf_train_dataset) * num_epochs
lr_scheduler = PolynomialDecay(
    initial_learning_rate=5e-5, end_learning_rate=0.0, decay_steps=num_train_steps
)
from tensorflow.keras.optimizers import Adam

opt = Adam(learning_rate=lr_scheduler)


# In[28]:


model = TFAutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)
loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer=opt, loss=loss, metrics=["accuracy"])


# In[29]:


start_tune = time.time()
model.fit(tf_train_dataset, validation_data=tf_validation_dataset, epochs=3)
end_tune = time.time() - start_tune
print("Fine Tuning Training Time: ", end_tune)


# In[30]:


preds = model.predict(tf_validation_dataset)["logits"]


# In[31]:


class_preds = np.argmax(preds, axis=1)
print(preds.shape, class_preds.shape)


# In[33]:


import evaluate

metric = evaluate.load("glue", "cola")
metric.compute(predictions=class_preds, references=dataset["validation"]["label"])


# In[ ]:




